core/src/main/java/com/linecorp/armeria/common/AbstractHttpRequestBuilder.java

```diff
@@ -411,42 +423,142 @@ private RequestHeaders requestHeaders() {
 
     private String buildPath() {
         checkState(path != null, "path must be set.");
-        final StringBuilder pathBuilder = new StringBuilder(path);
+
+        boolean hasQueryString = false;
         if (!disablePathParams) {
+            // Look for : or { first.
+            final int pathLen = path.length();
             int i = 0;
-            while (i < pathBuilder.length()) {
-                if (pathBuilder.charAt(i) == '{') {
-                    int j = i + 1;
-                    while (j < pathBuilder.length() && pathBuilder.charAt(j) != '}') {
-                        j++;
+            boolean hasPathParams = false;
+
+            loop: while (i < pathLen) {
+                switch (path.charAt(i)) {
+                    case ':':
+                        if (i + 1 < pathLen && path.charAt(i + 1) == '/') {
+                            if (i + 2 < pathLen && path.charAt(i + 2) == '/') {
+                                // Found '://', i.e. path is an absolute URI.
+                                final int pathStart = path.indexOf('/', i + 3);
+                                if (pathStart < 0) {
+                                    // The URI doesn't have a path, e.g. http://example.com:8080
+                                    i = pathLen;
+                                    break loop;
+                                } else {
+                                    i = pathStart + 1; // +1 to skip the first '/' in the path
+                                    continue;
+                                }
+                            } else {
+                                // Found ':/' - Skip.
+                                i += 2;
+                                continue;
+                            }
+                        } else {
+                            hasPathParams = true;
+                            break loop;
+                        }
+                    case '{':
+                        if (i + 1 < pathLen && path.charAt(i + 1) == '}') {
+                            // Found '{}' - Skip.
+                            i += 2;
+                            continue;
+                        } else {
+                            hasPathParams = true;
+                            break loop;
+                        }
+                    case '?':
+                        hasQueryString = true;
+                        break loop;
+                }
+                i++;
+            }
+
+            if (hasPathParams) {
+                // Replace path parameters.
+                final StringBuilder buf = new StringBuilder(pathLen + 32); // Add a little bit of wiggle room.
```

Comment:
Perhaps we can also use `TemporaryThreadLocals.get()` here because the builder is inflated when the length of `queryParams` is greater than 32 anyway?
Or we might add a variant such as `TemporaryThreadLocals.get().stringBuilder(int)`. 🤔

Reply:
We can't use a thread-local buffer here because to call `Object.toString()` in the concatenation loop. Cannot guarantee that `toString()` will use a thread-local buffer there again.

Reply:
Should we instead convert all path params upfront?

Reply:
Yeah, I think we can do that. Now using `TemporaryThreadLocals`.